File: day-5/day-5.py
# @author: Ezedin Fedlu
# @date: 2022-12-05
# @version: 1.0.0
# @description: Advent of Code 2022 - Day 5

import sys
from pathlib import Path
from typing import List

def parse_crates(f) -> List[List[str]]:
    lines = []
    while '1' not in (line := f.readline()):
        lines.append(line)
    n = int(line.strip().split()[-1])
    crates = [[] for _ in range(n)]
    while lines:
        line = lines.pop()
        for i in range(n):
            p = 1 + 4*i
            if p < len(line) and line[p] != ' ':
                crates[i].append(line[p])
    return crates


def parse_program(f) -> List[List[int]]:
    program = []
    while line := f.readline():
        _, k, _, fm, _, to = line.strip().split()
        program.append((int(k), int(fm), int(to)))
    return program


def part_1(crates, program) -> str:
    crates = [crate[:] for crate in crates]
    for k, fm, to in program:
        [crates[to - 1].append(crates[fm - 1].pop()) for _ in range(k)]
    return ''.join(crate[-1] for crate in crates)


def part_2(crates, program) -> str:
    crates = [crate[:] for crate in crates]
    for k, fm, to in program:
        tmp = [crates[fm-1].pop() for _ in range(k)]
        [crates[to-1].append(tmp.pop()) for _ in range(k)]
    return ''.join(crate[-1] for crate in crates)


def main():
    file_path = Path(sys.argv[1])
    with open(file_path, 'r') as f:
        crates = parse_crates(f)
        if line := f.readline().strip() != '':
            print('bad input')  # empty line
        program = parse_program(f)
        print(part_1(crates, program))
        print(part_2(crates, program))


if __name__ == '__main__':
    main()

File: day-5/day-5.rs
use std::fs::File;
use std::io::{BufRead, BufReader};


fn main() {
    let input_file_path = std::env::args().nth(1).unwrap();
    let file = File::open(input_file_path).unwrap();
    let reader = BufReader::new(file);
    let crates = parse_crates(reader.lines());
    let program = parse_program(reader.lines());
    println!("Part 1: {}", part_1(&crates, &program));
    println!("Part 2: {}", part_2(&crates, &program));
}

fn parse_crates(lns: &Vec<String>) -> Vec<Vec<char>> {
    let mut lines = Vec::new();
    for line in lns {
        if line.starts_with("1") {
            break;
        }
        lines.push(line.clone());
        line.clear();
    }
    let n = line.trim().split(' ').last().unwrap().parse::<usize>().unwrap();
    let mut crates = vec![vec![]; n];
    while let Some(line) = lines.pop() {
        for i in 0..n {
            let p = 1 + 4*i;
            if p < line.len() && line.chars().nth(p).unwrap() != ' ' {
                crates[i].push(line.chars().nth(p).unwrap());
            }
        }
    }
    crates.iter_mut().for_each(|crate_| crate_.reverse());
    crates
}

fn parse_program(lns: &Vec<String>) -> Vec<Vec<char>> {
    let mut program = vec![];
    let mut line = String::new();
    for line in lns {
        let mut iter = line.trim().split(' ');
        iter.next();
        let k = iter.next().unwrap().parse::<usize>().unwrap();
        iter.next();
        let fm = iter.next().unwrap().parse::<usize>().unwrap();
        iter.next();
        let to = iter.next().unwrap().parse::<usize>().unwrap();
        program.push((k, fm, to));
        line.clear();
    }
    program
}


fn part_1(crates: &Vec<Vec<char>>, program: &Vec<(usize, usize, usize)>) -> String {
    let mut crates = crates.iter().map(|crate_| crate_.clone()).collect::<Vec<Vec<char>>>();
    for (k, fm, to) in program {
        for _ in 0..*k {
            crates[*to - 1].push(crates[*fm - 1].pop().unwrap());
        }
    }

    crates.iter().map(|crate_| crate_.last().unwrap()).collect::<String>()
}

fn part_2(crates: &Vec<Vec<char>>, program: &Vec<(usize, usize, usize)>) -> String {
    let mut crates = crates.iter().map(|crate_| crate_.clone()).collect::<Vec<Vec<char>>>();
    for (k, fm, to) in program {
        let mut tmp = vec![];
        for _ in 0..*k {
            tmp.push(crates[*fm - 1].pop().unwrap());
        }
        for _ in 0..*k {
            crates[*to - 1].push(tmp.pop().unwrap());
        }

    }

    crates.iter().map(|crate_| crate_.last().unwrap()).collect::<String>()
}